File: ImGui - Prime' External/examples/example_win32_directx11/Auth/AuthSys.cpp
#include "AuthSys.hpp"
#include "json.hpp"
#include "wnetwrap.h"
#include <windows.h>

using json = nlohmann::json;

namespace AuthSys {

    api::api(std::string name, std::string ownerid, std::string secret, std::string version, std::string apiUrl) {
        this->name = name;
        this->ownerid = ownerid;
        this->secret = secret;
        this->version = version;
        this->apiUrl = apiUrl;
    }

    std::string api::GetHWID() {
        HW_PROFILE_INFO hwProfileInfo;
        if (GetCurrentHwProfile(&hwProfileInfo)) {
            return std::string(hwProfileInfo.szHwProfileGuid);
        }
        return "UNKNOWN_HWID";
    }

    std::string api::PostRequest(std::string endpoint, std::string jsonData) {
        std::string fullUrl = apiUrl;
        if (fullUrl.back() != '/') fullUrl += "/";
        fullUrl += "client/" + endpoint;

        wrap::Response r = wrap::HttpsRequest(
            wrap::Url{ fullUrl }, 
            wrap::Method{ "POST" }, 
            wrap::Body{ jsonData }, 
            wrap::Header{ {"Content-Type", "application/json"} }
        );
        return r.text;
    }

    void api::init() {
        std::string hwid = GetHWID();
        json j;
        j["app_secret"] = secret;
        j["version"] = version;
        j["hwid"] = hwid;
        j["app_name"] = name;

        std::string response = PostRequest("init", j.dump());
        
        try {
            json res = json::parse(response);
            if (res.contains("status") && 
                (res["status"] == "success" || res["status"] == "update_available")) {
                initialized = true;
                if (res.contains("variables") && res["variables"].is_object()) {
                    app_data.variables = res["variables"].dump();
                }
            } else {
                initialized = false;
            }
        } catch (...) {
            initialized = false;
        }
    }

    void api::login(std::string username, std::string password) {
        sessionid = "";
        if (!initialized) return;
        
        json j;
        j["app_secret"] = secret;
        j["username"] = username;
        j["password"] = password;
        j["hwid"] = GetHWID();
        j["session_length"] = 86400;

        std::string response = PostRequest("login", j.dump());
        
        try {
            json res = json::parse(response);
            if (res.contains("detail")) return;
            if (res.contains("success") && res["success"] == true) {
                if (res.contains("token")) {
                    sessionid = res["token"].get<std::string>();
                }
                user_data.username = username;
            }
        } catch (...) {
        }
    }

    void api::register_user(std::string username, std::string password, std::string license_key, std::string email) {
        if (!initialized) return;
        
        json j;
        j["app_secret"] = secret;
        j["username"] = username;
        j["password"] = password;
        j["license_key"] = license_key;
        j["hwid"] = GetHWID();
        if (!email.empty()) {
            j["email"] = email;
        }
        
        std::string response = PostRequest("register", j.dump());
    }

    void api::license(std::string key) {
        sessionid = "";
        if (!initialized) return;
        
        json j;
        j["app_secret"] = secret;
        j["license_key"] = key;
        j["hwid"] = GetHWID();
        j["session_length"] = 86400;

        std::string response = PostRequest("license_login", j.dump());
        
        try {
            json res = json::parse(response);
            if (res.contains("detail")) return;
            if (res.contains("success") && res["success"] == true) {
                if (res.contains("token")) {
                    sessionid = res["token"].get<std::string>();
                }
                if (res.contains("username")) {
                    user_data.username = res["username"].get<std::string>();
                }
            }
        } catch (...) {
        }
    }

    std::string api::var(std::string varName) {
        if (!initialized || app_data.variables.empty()) return "";
        try {
            json vars = json::parse(app_data.variables);
            if (vars.contains(varName)) {
                if (vars[varName].is_string()) {
                    return vars[varName].get<std::string>();
                }
                return vars[varName].dump();
            }
        } catch (...) {
        }
        return "";
    }
}

File: sdk/AuthSys-CSHARP-Example/Console/AuthSys.cs
using System;
using System.Diagnostics;
using System.Management;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using System.Collections.Generic;

namespace AuthSys
{
    public class api
    {
        public string name, ownerid, secret, version, apiUrl;
        public string sessionToken;
        public bool initialized;

        public class AppData
        {
            public JsonElement variables { get; set; }
        }

        public AppData app_data = new AppData();
        public JsonElement user_data;
        public string lastError = "";
        public JsonElement lastResponse;

        public class ResponseStructure
        {
            public string status { get; set; }
            public string message { get; set; }
            public string detail { get; set; }
            public bool success { get; set; }
            public string token { get; set; }
            public bool valid { get; set; }
            public string username { get; set; }
            public string email { get; set; }
            public string expires_at { get; set; }
            public JsonElement variables { get; set; }
            public JsonElement user { get; set; }
        }

        public ResponseStructure response = new ResponseStructure();
        private static readonly HttpClient client = new HttpClient { Timeout = TimeSpan.FromSeconds(30) };

        public api(string name, string ownerid, string secret, string version, string apiUrl = "https://authsys-main-production.up.railway.app/api/v1")
        {
            this.name = name;
            this.ownerid = ownerid;
            this.secret = secret;
            this.version = version;
            this.apiUrl = apiUrl.TrimEnd('/');
        }

        private string GetHWID()
        {
            try
            {
                string uuid = string.Empty;
                ManagementObjectSearcher searcher = new ManagementObjectSearcher("SELECT UUID FROM Win32_ComputerSystemProduct");
                foreach (ManagementObject wmi in searcher.Get())
                {
                    uuid = wmi["UUID"].ToString();
                }
                return uuid;
            }
            catch
            {
                return "UNKNOWN_HWID";
            }
        }

        private ResponseStructure PostRequest(string endpoint, object data, Dictionary<string, string> headers = null)
        {
            try
            {
                string json = JsonSerializer.Serialize(data);
                var content = new StringContent(json, Encoding.UTF8, "application/json");

                var request = new HttpRequestMessage(HttpMethod.Post, $"{apiUrl}/client/{endpoint}")
                {
                    Content = content
                };
                if (headers != null)
                {
                    foreach (var h in headers)
                        request.Headers.TryAddWithoutValidation(h.Key, h.Value);
                }

                var result = client.SendAsync(request).GetAwaiter().GetResult();
                var resultString = result.Content.ReadAsStringAsync().GetAwaiter().GetResult();

                var opts = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
                var parsed = JsonSerializer.Deserialize<ResponseStructure>(resultString, opts);
                response = parsed ?? new ResponseStructure();
                try { lastResponse = JsonDocument.Parse(resultString).RootElement.Clone(); } catch { }
                lastError = response.detail ?? "";
                return response;
            }
            catch (Exception ex)
            {
                lastError = ex.Message;
                return new ResponseStructure { success = false, detail = ex.Message };
            }
        }

        public void init()
        {
            lastError = "";
            initialized = false;

            var data = new
            {
                app_secret = this.secret,
                version = this.version,
                app_name = this.name,
                hwid = GetHWID()
            };

            response = PostRequest("init", data);

            if (response.status == "success" || response.status == "update_available")
            {
                initialized = true;
                app_data.variables = response.variables;
            }
            else
            {
                lastError = response.detail ?? response.message ?? "Init failed";
            }
        }

        public void register(string username, string password, string license_key, string email = "")
        {
            lastError = "";
            if (!initialized)
            {
                lastError = "init() failed or not called";
                return;
            }

            var data = new
            {
                app_secret = this.secret,
                username,
                password,
                license_key,
                email = string.IsNullOrEmpty(email) ? null : email,
                hwid = GetHWID()
            };

            response = PostRequest("register", data);

            if (!string.IsNullOrEmpty(response.detail))
                lastError = response.detail;
            else if (!response.success)
                lastError = "Registration failed";
        }

        public void login(string username, string password, int sessionLength = 86400)
        {
            lastError = "";
            sessionToken = null;
            if (!initialized)
            {
                lastError = "init() failed or not called";
                return;
            }

            var data = new
            {
                app_secret = this.secret,
                username,
                password,
                hwid = GetHWID(),
                session_length = sessionLength
            };

            response = PostRequest("login", data);

            if (!string.IsNullOrEmpty(response.detail))
                lastError = response.detail;
            else if (response.success && !string.IsNullOrEmpty(response.token))
            {
                sessionToken = response.token;
                user_data = response.user;
            }
            else if (!response.success)
                lastError = "Login failed: server returned success=false";
        }

        public void licenseLogin(string key, int sessionLength = 86400)
        {
            lastError = "";
            sessionToken = null;
            if (!initialized)
            {
                lastError = "init() failed or not called";
                return;
            }

            var data = new
            {
                app_secret = this.secret,
                license_key = key,
                hwid = GetHWID(),
                session_length = sessionLength
            };

            response = PostRequest("license-login", data);

            if (!string.IsNullOrEmpty(response.detail))
                lastError = response.detail;
            else if (response.success && !string.IsNullOrEmpty(response.token))
            {
                sessionToken = response.token;
                user_data = response.user;
            }
            else if (!response.success)
                lastError = "License login failed: server returned success=false";
        }

        public void licenseCheck(string key)
        {
            lastError = "";
            var data = new { app_secret = this.secret, license_key = key };
            response = PostRequest("license/check", data);
            if (!string.IsNullOrEmpty(response.detail)) lastError = response.detail;
        }

        public void verify()
        {
            lastError = "";
            if (string.IsNullOrEmpty(sessionToken))
            {
                lastError = "No active session. Login first.";
                return;
            }

            var headers = new Dictionary<string, string>
            {
                { "Authorization", $"Bearer {sessionToken}" },
                { "X-HWID", GetHWID() }
            };
            response = PostRequest("verify", null, headers);

            if (!string.IsNullOrEmpty(response.detail))
                lastError = response.detail;
            else if (!response.valid)
                lastError = "Session verification failed";
        }

        public void chatSend(int roomId, string message)
        {
            lastError = "";
            if (string.IsNullOrEmpty(sessionToken))
            {
                lastError = "No active session. Login first.";
                return;
            }

            var headers = new Dictionary<string, string>
            {
                { "Authorization", $"Bearer {sessionToken}" }
            };
            response = PostRequest($"chat/send?room_id={roomId}&message={Uri.EscapeDataString(message)}", null, headers);
            if (!string.IsNullOrEmpty(response.detail)) lastError = response.detail;
        }

        public string var(string varName)
        {
            if (!initialized || app_data.variables.ValueKind == JsonValueKind.Undefined) return null;
            if (app_data.variables.TryGetProperty(varName, out JsonElement el))
                return el.GetString();
            return null;
        }

        public void logout()
        {
            sessionToken = null;
            user_data = default;
        }
    }
}
